internal/packages/walk.go

```diff
@@ -415,3 +422,38 @@ func excludedForSubdir(excluded []string, subdir string) []string {
 	}
 	return filtered
 }
+
+type symlinkResolver struct {
+	root    string
+	visited []string
+}
+
+// Decide if symlink dir/base should be followed.
+func (r *symlinkResolver) follow(dir, base string) bool {
+	if dir == r.root && strings.HasPrefix(base, "bazel-") {
+		// Links such as bazel-<workspace>, bazel-out, bazel-genfiles are created by
+		// Bazel to point to internal build directories.
+		return false
+	}
+	// See if the symlink points to a tree that has been already visited.
+	fullpath := filepath.Join(dir, base)
+	dest, err := os.Readlink(fullpath)
+	if err != nil {
+		log.Printf("readlink %v: %v", fullpath, err)
+		return false
+	}
+	dest = filepath.Join(dir, dest)
+	for _, p := range r.visited {
```

Comment:
Use `pathutils.HasPrefix(dest, r.root)` to check whether the link is inside the workspace. `pathutils` is an internal package that is aware of path components. `r.root` will already be a canonical path.

To check whether a directory has been visited, it would be faster to use a `map` (since we won't need to check prefixes anymore).

Reply:
I thought about using a map, but file.Info doesn't offer a good cross-platform key.